chain-tx-enclave-next/mls/src/tree.rs

```diff
@@ -335,34 +441,264 @@ impl Tree {
+            self.nodes[node_index.0]
+                .parent_node_mut()
+                .expect("implant invalid node")
+                .set_secret(self.cs, &last_secret);
+        }
+
+        // Define `commit_secret` as the value `path_secret[n+1]` derived from the
+        // `path_secret[n]` value assigned to the root node.
+        let commit_secret = self
+            .cs
+            .expand_label(&last_secret, vec![], "path", &[], self.cs.secret_size())
+            .expect("invalid length");
+
+        // update parent hash
+        let leaf_parent_hash = self.set_parent_hash_path(sender);
+
+        (leaf_parent_hash, commit_secret)
+    }
+
+    fn set_parent_hash_path(&mut self, sender: LeafSize) -> Vec<u8> {
+        let from = NodeSize::from_leaf_index(sender);
+        let path = from.direct_path(self.leaf_len());
+        for (node, parent) in path.iter().copied().zip(path.iter().skip(1).copied()).rev() {
+            // no leaf node
+            let hash = self.nodes[parent.0].compute_parent_hash(self.cs);
+            if let Some(n) = self.nodes[node.0].parent_node_mut() {
+                n.parent_hash = hash;
+            }
+        }
+        self.nodes[path[0].0].compute_parent_hash(self.cs)
+    }
 }
 
 fn node_hash(nodes: &[Node], cs: CipherSuite, index: NodeSize, leaf_size: LeafSize) -> Vec<u8> {
     let node = &nodes[index.0];
-    match node {
-        Node::Leaf(kp) => {
-            let mut inp = Vec::new();
-            LeafNodeHashInput {
-                node_index: index.0 as u32,
-                key_package: kp.clone(),
-            }
-            .encode(&mut inp);
-            cs.hash(&inp)
+    let payload = match node {
+        Node::Leaf(kp) => LeafNodeHashInput {
+            node_index: index.0 as u32,
+            key_package: kp.clone(),
         }
+        .get_encoding(),
         Node::Parent(pn) => {
-            let left_index = index.left().unwrap();
+            let left_index = index.left().expect("parent node won't panic");
             let left_hash = node_hash(nodes, cs, left_index, leaf_size);
-            let right_index = index.right(leaf_size).unwrap();
+            let right_index = index.right(leaf_size).expect("parent node won't panic");
             let right_hash = node_hash(nodes, cs, right_index, leaf_size);
-            cs.hash(
-                &ParentNodeHashInput {
-                    node_index: index.0 as u32,
-                    parent_node: pn.clone(),
-                    left_hash,
-                    right_hash,
-                }
-                .get_encoding(),
-            )
+            ParentNodeHashInput {
+                node_index: index.0 as u32,
+                parent_node: pn.clone(),
+                left_hash,
+                right_hash,
+            }
+            .get_encoding()
         }
+    };
+    cs.hash(&payload)
+}
+
+/// no blank nodes return
+fn resolve(nodes: &[Node], index: NodeSize) -> Vec<NodeSize> {
```

Comment:
could be good to link "Ratchet Tree Nodes" section -- I was trying to search "resolve" where it is defined

Reply:
done